### PR TITLE
Add onBytesReceived callback to consolidateHttpClientResponseBytes()

### PR DESCRIPTION
**Pull request by tvolkert**
## Description

This will allow us to plumb the byte chunks in a chunked
network response up to the higher levels of the framework
to notify interested parties of network loading progress.

## Related Issues

https://github.com/flutter/flutter/issues/32374

## Tests

I added the following tests:

* Verification that the `onBytesReceived` callback is invoked correctly.
* Verification that the method handles the gzip permutations correctly.
* Verification that if the callback throws, the error propagates to the future that is returned by the function.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by Hixie** [comment]
cc @goderbauer

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/foundation/consolidate_response.dart [hunk 1]
```diff
@@ -3,31 +3,121 @@
...
+/// consolidating the bytes of an [HttpClientResponse] into a [Uint8List].
 ///
-/// The future returned will forward all errors emitted by [response].
-Future<Uint8List> consolidateHttpClientResponseBytes(HttpClientResponse response) {
-  // response.contentLength is not trustworthy when GZIP is involved
-  // or other cases where an intermediate transformer has been applied
-  // to the stream.
+/// The `cumulative` parameter will contain the total number of bytes received
+/// thus far. If the response has been gzipped, this number will be the number
+/// of compressed bytes that have received _across the wire_.
```
[the remark this answers is not in the document]
Comment (tvolkert) [reply packages/flutter/lib/src/foundation/consolidate_response.dart]: done

File: packages/flutter/lib/src/foundation/consolidate_response.dart [hunk 2]
```diff
@@ -3,31 +3,121 @@
...
+/// returned to the client and the total size of the response may not be known
+/// until the request has been fully processed).
+///
+/// This is used in [consolidateHttpClientResponseBytes].
+typedef BytesReceivedCallback = void Function(int cumulative, int total);
+
+/// Efficiently converts the response body of an [HttpClientResponse] into a
+/// [Uint8List].
+///
+/// The future returned will forward any error emitted by [response].
```
[the remark this answers is not in the document]
Comment (tvolkert) [reply packages/flutter/lib/src/foundation/consolidate_response.dart]: done

**Pull request merged**